rust-runtime/smithy-eventstream/src/frame.rs

```diff
@@ -136,9 +143,18 @@ mod value {
                     buffer.put_u16(checked(val.as_bytes().len(), Error::HeaderValueTooLong)?);
                     buffer.put_slice(&val.as_bytes()[..]);
                 }
-                Timestamp(val) => {
+                Timestamp(time) => {
                     buffer.put_u8(TYPE_TIMESTAMP);
-                    buffer.put_i64(val.epoch_seconds());
+                    buffer.put_i64(
+                        time.epoch_seconds()
+                            .checked_mul(1000)
+                            .and_then(|val| {
+                                val.checked_add(
+                                    i64::from(time.subsec_nanos()) / NANOS_PER_MILLI_I64,
+                                )
+                            })
+                            .ok_or(Error::TimestampValueTooLarge(*time))?,
```

Comment:
I wonder if we should just remove the support for negative times. I _think_ this actually wrong for negative values of epoch seconds because `subsec_nanos()` is always positive. _Technically_ you should be subtracting the subsecond nanos if epoch seconds is negative.

Which makes me think we should add `time.as_millis() -> i64` to instant that handles this properly.

Reply:
(and as above, I'd go through duration instead of doing all the multiplying / dividing by hand)